backtester/engine/live.go

```diff
@@ -2,138 +2,470 @@ package engine
+
+	results := make([]bool, len(d.sourcesToCheck))
+	// timeToRetrieve ensures consistent data retrieval
+	// in the event of a candle rollover mid-loop
+	timeToRetrieve := time.Now()
+	for i := range d.sourcesToCheck {
+		if d.verboseDataCheck {
+			log.Infof(common.LiveStrategy, "%v %v %v checking for new data", d.sourcesToCheck[i].exchangeName, d.sourcesToCheck[i].asset, d.sourcesToCheck[i].pair)
+		}
+		var updated bool
+		updated, err = d.sourcesToCheck[i].loadCandleData(timeToRetrieve)
+		if err != nil {
+			return false, err
+		}
+		results[i] = updated
+	}
+	allUpdated := true
+	for i := range results {
+		if !results[i] {
+			allUpdated = false
+		}
+	}
+	if !allUpdated {
+		return false, nil
+	}
+	for i := range d.sourcesToCheck {
+		if d.verboseDataCheck {
+			log.Infof(common.LiveStrategy, "%v %v %v found new data", d.sourcesToCheck[i].exchangeName, d.sourcesToCheck[i].asset, d.sourcesToCheck[i].pair)
+		}
+		d.sourcesToCheck[i].pairCandles.AppendResults(d.sourcesToCheck[i].candlesToAppend)
+		d.sourcesToCheck[i].candlesToAppend.Candles = nil
+		d.dataHolder.SetDataForCurrency(d.sourcesToCheck[i].exchangeName, d.sourcesToCheck[i].asset, d.sourcesToCheck[i].pair, &d.sourcesToCheck[i].pairCandles)
+		err = d.report.SetKlineData(&d.sourcesToCheck[i].pairCandles.Item)
+		if err != nil {
+			log.Errorf(common.LiveStrategy, "%v %v %v issue processing kline data: %v", d.sourcesToCheck[i].exchangeName, d.sourcesToCheck[i].asset, d.sourcesToCheck[i].pair, err)
+		}
+		err = d.funding.AddUSDTrackingData(&d.sourcesToCheck[i].pairCandles)
+		if err != nil && !errors.Is(err, funding.ErrUSDTrackingDisabled) {
+			log.Errorf(common.LiveStrategy, "%v %v %v issue processing USD tracking data: %v", d.sourcesToCheck[i].exchangeName, d.sourcesToCheck[i].asset, d.sourcesToCheck[i].pair, err)
+		}
+	}
+	if !d.hasUpdatedFunding {
+		err = d.UpdateFunding(false)
+		if err != nil {
 			if err != nil {
-				log.Error(common.Backtester, err)
-				return
+				log.Error(common.LiveStrategy, err)
 			}
 		}
 	}
+
+	return true, nil
 }
 
-func (bt *BackTest) loadLiveData(resp *kline.DataFromKline, cfg *config.Config, exch gctexchange.IBotExchange, fPair currency.Pair, a asset.Item, dataType int64) error {
-	if resp == nil {
-		return errNilData
+// SetDataForClosingAllPositions is triggered on a live data run
+// when closing all positions on close is true.
+// it will ensure all data is set such as USD tracking data
+func (d *dataChecker) SetDataForClosingAllPositions(s ...signal.Event) error {
+	if d == nil {
+		return fmt.Errorf("%w dataChecker", gctcommon.ErrNilPointer)
 	}
-	if cfg == nil {
-		return errNilConfig
+	if len(s) == 0 {
+		return fmt.Errorf("%w signal events", gctcommon.ErrNilPointer)
 	}
-	if exch == nil {
-		return errNilExchange
+	d.m.Lock()
+	defer d.m.Unlock()
+	var err error
+
+	setData := false
+	for x := range s {
+		if s[x] == nil {
+			return fmt.Errorf("%w signal events", errNilData)
+		}
+		for y := range d.sourcesToCheck {
```

Comment:
I'm not so sure on that, since it constantly will for loop to check each data source. For  loops on maps are slow. Applying via map is certainly faster though